src/Oscoin/Protocol/Sync.hs

```diff
@@ -0,0 +1,393 @@
+    deriving (Show, Eq)
+
+-- | The singleton type for 'ProtocolRequest', capable of carrying some
+-- witness (in this case, 'ProtocolRequest').
+data SProtocolRequest r where
+  SGetTip :: SProtocolRequest 'GetTip
+  SGetBlocks :: SProtocolRequest 'GetBlocks
+  SGetBlockHeaders :: SProtocolRequest 'GetBlockHeaders
+
+-- | A closed type family mapping each 'ProtocolRequest' to its list of
+-- arguments, so that we can use the 'ProtocolRequest' (where all constructors
+-- have kind /*/) as a constraint.
+type family ProtocolRequestArgs (c :: ProtocolRequest) :: * where
+    ProtocolRequestArgs 'GetTip          = ()
+    ProtocolRequestArgs 'GetBlocks       = Range
+    ProtocolRequestArgs 'GetBlockHeaders = Range
+
+-- | A protocol response, indexed by the corresponding 'ProtocolRequest, so
+-- that matching the wrong type will result in a type error.
+type family ProtocolResponse c tx s (r :: ProtocolRequest) :: *
+
+type instance ProtocolResponse c (Tx c) PoW 'GetTip =
+        Block c (Tx c) (Sealed c PoW)
+type instance ProtocolResponse c (Tx c) PoW 'GetBlocks =
+        OldestFirst [] (Block c (Tx c) (Sealed c PoW))
+type instance ProtocolResponse c tx PoW 'GetBlockHeaders =
+        OldestFirst [] (BlockHeader c (Sealed c PoW))
+
+-- | A data fetcher is a wrapper around an action that given a (singleton)
+-- 'SProtocolRequest' and a set of arguments, returns a 'ProtocolResponse'
+-- while producting some side-effect in @m@.
+newtype DataFetcher c tx s m =
+    DataFetcher {
+      fetch :: forall r. ActivePeer c
+            -> SProtocolRequest r
+            -> ProtocolRequestArgs r
+            -> m (Either Timeout (ProtocolResponse c tx s r))
+                }
+
+-- | An opaque reference to a record of functions which can be used to query
+-- the network and the other parts of the system for information.
+data SyncContext c tx s m = SyncContext
+    { scNu               :: Nu
+    -- ^ Used to determine the block tolerance to consider
+    -- ourselves synced. It correspond to /nu/ from the spec, and is used
+    -- in the calculation of 'isDone'.
+    , scActivePeers       :: m (ActivePeers c)
+    -- ^ An effectful action that can be used to get the set of active peers
+    -- for this node.
+    , scDataFetcher       :: DataFetcher c tx s m
+    -- ^ A 'DataFetcher'.
+    , scEventTracer       :: Tracer m
+    -- ^ A tracer for interesting events that happens during syncing.
+    , scConcurrently      :: forall a b t.  Traversable t => t a -> (a -> m b) -> m (t b)
+    -- ^ An action that can be used to fetch some resources in parallel.
+    -- 'IO'-based implementations will probably use 'forConcurrently' or some
+    -- variation of it.
+    , scLocalChainReader  :: BlockStoreReader c tx s m
+    -- ^ A read-only handle to the local blockchain.
+    , scUpstreamConsumers :: [SyncEvent c tx s -> m ()]
+    -- ^ A list of \"upstream consumers\", represented as effectful actions from
+    -- a 'SyncEvent' to 'm ()'.
```

Comment:
Based on the type, this is more a list of "event handlers", which is much more powerful.